### ISSUE TITLE
nixos: alsa-store.service crashes because `/var/lib/alsa/asound.state` does not exist

### ISSUE THREAD

**Issue by Atemu** [NixOS/nixpkgs#319809]
### Describe the bug

On my new laptop with a brand new installation, I run into this service failing to start: 

```
Jun 14 12:37:40 LYKOURGOS systemd[1]: Starting Store Sound Card State...
Jun 14 12:37:40 LYKOURGOS alsactl[73843]: /nix/store/xqxc5fivgzbwqwg6nyw96imanvkv36nh-alsa-utils-1.2.10/sbin/alsactl: load_state:1689: Cannot open /var/lib/alsa/asound.state for reading: No such file or directory
Jun 14 12:37:40 LYKOURGOS alsactl[73843]: Found hardware: "HDA-Intel" "ATI R6xx HDMI" "HDA:1002aa01,00aa0100,00100800" "0xf111" "0x0005"
Jun 14 12:37:40 LYKOURGOS alsactl[73843]: Hardware is initialized using a generic method
Jun 14 12:37:40 LYKOURGOS alsactl[73843]: Found hardware: "HDA-Intel" "Realtek ALC295" "HDA:10ec0295,f1110005,00100002" "0xf111" "0x0005"
Jun 14 12:37:40 LYKOURGOS alsactl[73843]: Hardware is initialized using a generic method
Jun 14 12:37:40 LYKOURGOS systemd[1]: alsa-store.service: Main process exited, code=exited, status=99/n/a
Jun 14 12:37:40 LYKOURGOS systemd[1]: alsa-store.service: Failed with result 'exit-code'.
Jun 14 12:37:40 LYKOURGOS systemd[1]: Failed to start Store Sound Card State.
```

On my desktop (whose installation dates back to 20.09), this file exists.

### Steps To Reproduce

Unknown. I appear to be first person to run into this bug. I merely installed NixOS with sound enabled via pipewire and this is what I get.

My config is relatively vanilla NixOS; no external projects. Key points I think are relevant are that I use pipewire with alsa compat and that I installed very recently using 7d916e720af6b2ca355e4d0cfb8e4f742c172239.

### Expected behavior

It should not fail to start.

Should the file exist though? What is supposed to create it?

### Additional context

Is this service even relevant with pipewire?

### Notify maintainers

@Kranzes @K900 (pipewire maintainers)

@edolstra (added alsa-store.service) 

@comfysage (authored 4b0fd3799011f074c15531bba6027b04133e4e10 which seems related)

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
output here
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by K900** [comment]
We should really nuke that stupid service. Just disable `sound.enable`, it doesn't actually do anything useful.

**Comment by comfysage** [comment]
> Should the file exist though? What is supposed to create it?

when the service starts it's supposed to look for the file and source it. if it does not exist it should NOT error since the flag `--ignore` is given.
the file is created by the `ExecStop` field which stores the current state in the file (on shutdown).

**Comment by eclairevoyant** [comment]
@D3vil0p3r did you remove `sound.enable`? that is not meant to be used with pipewire, pw has its own alsa-compat

**Comment by eclairevoyant** [comment]
Yeah, the `services.pipewire.alsa.*` are there for alsa compat, if you need it. (Personally I haven't found an ALSA-only app in my entire year using wayland, but the options are there if needed.)

**Comment by eclairevoyant** [comment]
Created #319839 to address this confusion

**Comment by damccull** [comment]
I just ran into this after not having the issue for a couple months. Don't really know what changed. But I didn't add sound.enable myself, I don't think. I'm pretty certain the installer generated it along with the pipewire config. Is this something that needs to be looked at as well?

**Comment by eclairevoyant** [comment]
All the installer scripts have `sound.enable` removed.

**Comment by damccull** [comment]
Excellent. Hmm, maybe I pasted it in at some point by accident. Thanks.

**Comment by Fuuzetsu** [comment]
I updated my NixOS to very recent master and now when it tries to switch over, I get

```
stopping the following units: accounts-daemon.service
activating the configuration...
setting up /etc...
reloading user units for shana...
restarting sysinit-reactivation.target
reloading the following units: dbus.service
restarting the following units: polkit.service
starting the following units: accounts-daemon.service
the following new units were started: libvirtd.service
warning: the following units failed: alsa-store.service

× alsa-store.service - Store Sound Card State
     Loaded: loaded (/etc/systemd/system/alsa-store.service; enabled; preset: enabled)
     Active: failed (Result: exit-code) since Thu 2024-07-11 16:07:08 JST; 136ms ago
    Process: 249378 ExecStartPre=/nix/store/dhv5gh89him9a7ddr56cqg87zfkmjihp-coreutils-9.5/bin/mkdir -p /var/lib/alsa (code=exited, status=0/SUCCESS)
    Process: 249382 ExecStart=/nix/store/kkca6yxi479ji5zirgvkbgas94bgksc3-alsa-utils-1.2.10/sbin/alsactl restore --ignore (code=exited, status=99)
   Main PID: 249382 (code=exited, status=99)
         IP: 0B in, 0B out
        CPU: 8ms

Jul 11 16:07:08 aya alsactl[249382]: /nix/store/kkca6yxi479ji5zirgvkbgas94bgksc3-alsa-utils-1.2.10/sbin/alsactl: load_state:1689: Cannot open /var/lib/alsa/asound.state for reading: No such file or directory
Jul 11 16:07:08 aya alsactl[249382]: Found hardware: "HDA-Intel" "Realtek ALC1220" "HDA:10ec1220,1458a0d5,00100101 HDA:80862818,80860101,00100000" "0x1458" "0xa0d5"
Jul 11 16:07:08 aya alsactl[249382]: Hardware is initialized using a generic method
Jul 11 16:07:08 aya alsactl[249382]: Found hardware: "HDA-Intel" "Nvidia GPU 82 HDMI/DP" "HDA:10de0082,14623607,00100100" "0x1462" "0x3607"
Jul 11 16:07:08 aya alsactl[249382]: Hardware is initialized using a generic method
Jul 11 16:07:08 aya alsactl[249382]: Found hardware: "USB-Audio" "USB Mixer" "USBb58e:9e84" "" ""
Jul 11 16:07:08 aya alsactl[249382]: Hardware is initialized using a generic method
Jul 11 16:07:08 aya systemd[1]: alsa-store.service: Main process exited, code=exited, status=99/n/a
Jul 11 16:07:08 aya systemd[1]: alsa-store.service: Failed with result 'exit-code'.
Jul 11 16:07:08 aya systemd[1]: Failed to start Store Sound Card State.
warning: error(s) occurred while switching to the new configuration
```

I don't know if it's the same issue as I don't have anything about alsa or pipewire or anything even mentioning `sound` in my configuration.nix...

**Comment by dbaynard** [comment]
@Fuuzetsu I had the same experience. Following advice from the wiki/manual I'd set `sound.enable` to `true` and so removing this line removed the (crashing) service, while having no effect on sound output.

**Comment by K900** [comment]
That option name is extremely misleading, see https://github.com/NixOS/nixpkgs/pull/326262.

**Comment by dbaynard** [comment]
Ah thanks, I'd subscribed to the previous PR but I hadn't noticed it had been closed without a merge.

**Comment by Fuuzetsu** [comment]
> @Fuuzetsu I had the same experience. Following advice from the wiki/manual I'd set `sound.enable` to `true` and so removing this line removed the (crashing) service, while having no effect on sound output.

I was trying to say that I haven't hsd this enabled in the first place and merely updating nixpkgs has started to cause the error. Maybe it's a different problem with the same symptom, unsure.

**Comment by K900** [comment]
`services.pulseaudio.enable` enables `sound.enable` as well.

**Comment by Fuuzetsu** [comment]
FWIW after updating to current master, the problem has gone away for me

**Comment by FloralExMachina** [comment]
I recently fixed this by running `sudo alsactl store --ignore` manually and restarted the service

**Comment by hacker1024** [comment]
@comfysage
> when the service starts it's supposed to look for the file and source it. if it does not exist it should NOT error since the flag --ignore is given.

The --ignore flag is documented as "ignore 'No soundcards found' error". This is a different error to a missing `/var/lib/alsa/asound.state` file (which is the case when no store has yet been performed).

It looks like the proper solution here is to check that that file exists before running the restore command.

**Comment by comfysage** [comment]
@hacker1024 

> The --ignore flag is documented as "ignore 'No soundcards found' error". This is a different error to a missing `/var/lib/alsa/asound.state` file (which is the case when no store has yet been performed).
> 
> It looks like the proper solution here is to check that that file exists before running the restore command.

the issue was fixed in https://github.com/NixOS/nixpkgs/commit/c8a84a01d57ea72a5e560f69cdd8596b7f7421b2 where it was decided to (instead of checking whether the file exist) ignore the error code when the store file isnt found.
